Sources/Build/BuildOperation.swift

```diff
@@ -530,33 +529,44 @@ public final class BuildOperation: PackageStructureDelegate, SPMBuildCore.BuildS
-                    let handledFiles = result.flatMap { $0.buildCommands.flatMap { $0.inputFiles } }
-                    unhandledFiles.subtract(handledFiles)
-                }
-                if unhandledFiles.isEmpty { continue }
-
-                // Emit a diagnostic if any remain. This is kept the same as the previous message for now, but this could be improved.
-                let diagnosticsEmitter = self.observabilityScope.makeDiagnosticsEmitter {
-                    var metadata = ObservabilityMetadata()
-                    metadata.packageIdentity = package.identity
-                    metadata.packageKind = package.manifest.packageKind
-                    metadata.targetName = target.name
-                    return metadata
-                }
-                var warning = "found \(unhandledFiles.count) file(s) which are unhandled; explicitly declare them as resources or exclude from the target\n"
-                for file in unhandledFiles {
-                    warning += "    " + file.pathString + "\n"
-                }
-                diagnosticsEmitter.emit(warning: warning)
+        // rdar://113256834 This fix works for the plugins that do not have PreBuildCommands.
+        let targetsToConsider: [ResolvedTarget]
+        if let subset = subset, let recursiveDependencies = try 
+            subset.recursiveDependencies(for: graph, observabilityScope: observabilityScope) {
+            targetsToConsider = recursiveDependencies
+        } else {
+            targetsToConsider = Array(graph.reachableTargets)
+        }
+
+        for target in targetsToConsider {
+            guard let package = graph.package(for: target), package.manifest.toolsVersion >= .v5_3 else {
+                continue
             }
+
+            // Get the set of unhandled files in targets.
+            var unhandledFiles = Set(target.underlying.others)
+            if unhandledFiles.isEmpty { continue }
+
+            // Subtract out any that were inputs to any commands generated by plugins.
+            if let result = buildToolPluginInvocationResults[target.id]?.results {
+                let handledFiles = result.flatMap{ $0.buildCommands.flatMap{ $0.inputFiles } }
+                unhandledFiles.subtract(handledFiles)
+            }
+            if unhandledFiles.isEmpty { continue }
+
+            // Emit a diagnostic if any remain. This is kept the same as the previous message for now, but this could be improved.
+            let diagnosticsEmitter = self.observabilityScope.makeDiagnosticsEmitter {
+                var metadata = ObservabilityMetadata()
+                metadata.packageIdentity = package.identity
+                metadata.packageKind = package.manifest.packageKind
+                metadata.targetName = target.name
+                return metadata
+            }
+            var warning = "found \(unhandledFiles.count) file(s) which are unhandled; explicitly declare them as resources or exclude from the target\n"
```

Comment:
Would a multiline string work here? It would also make line length warnings for formatters and linters go away.
```suggestion
            var warning = """
                found \(unhandledFiles.count) file(s) which are unhandled; explicitly declare them \
                as resources or exclude from the target\n
                """
```